043-tasks_queue/proposal.md

```diff
@@ -0,0 +1,145 @@
+    chosenWorker, err := client.chooseTaskWorker(...)
+    //...
+}
+```
+This function is called from the `Run()` method of TaskStep, which is effectively responsible for starting the task's container.
+
+The task chooses a worker given the placement strategy and the worker starts executing the corresponding container. This happens without taking into account the available resources on the worker. If users submit more jobs than the workers can instantaneously handle, the system becomes overloaded. Additionally, in case of overload no corrective action is possible since the tasks cannot re-schedule themselves.
+
+Instead of directly allocating a worker from a task, we propose the following, extended worker allocation process:
+
+![resource pool schema](https://ci-pix4d-internal-static.s3-eu-west-1.amazonaws.com/concourse-scheduling-rfc.png)
+
+A new component, called Resource Pool is placed between the worker client (the ATC) and worker pool. The client would interact with the worker pool as follows:
+
+1. The client requests a worker from Resource Pool.  This call blocks until a worker becomes free.
+2. The resource pool returns the worker to the client.  The worker has been chosen from the available workers using the container placement strategy.
+3. The client starts the task container on the allocated worker
+4. The task container stops
+5. The worker client releases the worker's resources
+
+The key idea is the resource pool throttles the starting of containers and protects the workers from overload.
+
+In this schema the resource pool only increases and decreases the number of active tasks on the workers.  In the next section we show that the proposal is not specific to the active tasks. The resource pool could be aware of other parameters that describe the workers' available capacity.
+
+## Resource Pool
+
+The new ResourcePool component has the following responsibilities:
+
+* observe the number of active tasks on the worker
+* offer workers' resources to tasks
+
+Concretely, we propose the following interface
+```go
+type Resources = map[string]int  // for example: {"active-task": 1}
+
+type ResourcePool interface {
+    RequestWorker(requiredResources Resources, strategy PlacementStrategy, task parameters...) -> Worker
+    ReleaseWorker(requiredResources Resources, worker Worker)
+}
+```
+
+This interface mimics the fundamental operations of locks and semaphores.
+
+Now, the interaction between the Client and the ResourcePool integrates unobtrusively into the current code base. Instead of tasks sending themselves immediately to the workers, the same part of the code could be, schematically
+
+```go
+func (client *client) RunTaskStep(...) TaskResult {
+    //...
+
+    requiredResources := make(map[string]int)
+    requiredResources["active-task"] = 1
+ 
+    chosenWorker := resourcePool.RequestWorker(requiredResoures, placementStrategy, taskParameters...)
+    defer resourcePool.ReleaseWorker(requiredREsources, chosenWorker)
```

Comment:
Although this is captured as a `defer`, it's probably worth reasoning about the doomsday scenario of an ATC getting `kill -9`ed. Would these requests leak? Do we need some sort of GC semantics for them? 🤔

Reply:
IIRC when the ATC starts again (or a different ATC for that matter), it will go through this code again, then if the task is still being executed it will wait for it, otherwise will exit immediately, calling the defer.

This is already the case for the placement strategy `limit-active-tasks` where we indeed tested that when the ATC recovers all the "flying" tasks also releases the resources no longer in use.